management/server/user.go

```diff
@@ -472,56 +502,106 @@ func (am *DefaultAccountManager) SaveUser(accountID, userID string, update *User
 	if oldUser == nil {
-		return nil, status.Errorf(status.NotFound, "update not found")
+		return nil, status.Errorf(status.NotFound, "user to update doesn't exist")
+	}
+
+	if initiatorUser.IsAdmin() && initiatorUserID == update.Id && oldUser.Blocked != update.Blocked {
+		return nil, status.Errorf(status.PermissionDenied, "admins can't block or unblock themselves")
+	}
+
+	if initiatorUser.IsAdmin() && initiatorUserID == update.Id && update.Role != UserRoleAdmin {
+		return nil, status.Errorf(status.PermissionDenied, "admins can't change their role")
 	}
 
 	// only auto groups, revoked status, and name can be updated for now
 	newUser := oldUser.Copy()
 	newUser.AutoGroups = update.AutoGroups
 	newUser.Role = update.Role
+	newUser.Blocked = update.Blocked
+
+	for _, newGroupID := range update.AutoGroups {
+		if _, ok := account.Groups[newGroupID]; !ok {
+			return nil, status.Errorf(status.InvalidArgument, "provided group ID %s in the user %s update doesn't exist",
+				newGroupID, update.Id)
+		}
+	}
+	newUser.AutoGroups = update.AutoGroups
 
 	account.Users[newUser.Id] = newUser
 
+	if oldUser.Blocked != update.Blocked {
+		// expire peers that belong to the user who's getting blocked
+		if update.Blocked {
+			blockedPeers, err := account.FindUserPeers(update.Id)
+			if err != nil {
+				return nil, err
+			}
+			var peerIDs []string
+			for _, peer := range blockedPeers {
+				peerIDs = append(peerIDs, peer.ID)
+				peer.MarkLoginExpired(true)
+				account.UpdatePeer(peer)
+				err = am.Store.SavePeerStatus(account.Id, peer.ID, *peer.Status)
+				if err != nil {
+					log.Errorf("failed saving peer status while expiring peer %s", peer.ID)
+					return nil, err
+				}
+			}
+			am.peersUpdateManager.CloseChannels(peerIDs)
+			err = am.updateAccountPeers(account)
+			if err != nil {
+				log.Errorf("failed updating account peers while expiring peers of a blocked user %s", accountID)
+				return nil, err
+			}
+		}
+	}
+
 	if err = am.Store.SaveAccount(account); err != nil {
 		return nil, err
 	}
 
 	defer func() {
+		// store activity logs
 		if oldUser.Role != newUser.Role {
-			am.storeEvent(userID, oldUser.Id, accountID, activity.UserRoleUpdated, map[string]any{"role": newUser.Role})
+			am.storeEvent(initiatorUserID, oldUser.Id, accountID, activity.UserRoleUpdated, map[string]any{"role": newUser.Role})
 		}
 
-		removedGroups := difference(oldUser.AutoGroups, update.AutoGroups)
-		addedGroups := difference(newUser.AutoGroups, oldUser.AutoGroups)
-		for _, g := range removedGroups {
-			group := account.GetGroup(g)
-			if group != nil {
-				am.storeEvent(userID, oldUser.Id, accountID, activity.GroupRemovedFromUser,
-					map[string]any{"group": group.Name, "group_id": group.ID, "is_service_user": newUser.IsServiceUser, "user_name": newUser.ServiceUserName})
-			} else {
-				log.Errorf("group %s not found while saving user activity event of account %s", g, account.Id)
-			}
+		if update.AutoGroups != nil {
```

Comment:
I think Autogroups should never be nil?? If it can be are we sure nil can not mean empty list of groups? because then the remove part might be needed